app/src/main/java/ch/hikemate/app/model/elevation/ElevationRepositoryCopernicus.kt

```diff
@@ -0,0 +1,414 @@
+
+    // Limit the cache size to avoid memory issues
+    // This has been tested with a cache size of 150000, on a Google Pixel 8 Pro, without any issues
+    private const val DEFAULT_CACHE_MAX_SIZE = 100000
+
+    // The maximum number of failed requests before giving up
+    private const val MAX_FAILED_REQUESTS = 5
+
+    // Delay to add per failed request
+    private const val FAILED_REQUEST_DELAY = 1000L
+
+    // The max number of coordinates that can be sent in a single request
+    const val MAX_COORDINATES_PER_REQUEST = 500
+
+    // The log tag
+    private val LOG_TAG = this::class.simpleName
+  }
+
+  // Cache for the elevation data, indexed by hike ID
+  private val cache = hashMapOf<LatLong, ElevationCacheEntry>()
+
+  // List of requests that are waiting for the cache to be updated
+  private val requests = mutableListOf<ElevationRequest>()
+
+  // A mutex to ensure that the cache and requests are accessed atomically
+  private val mutex: Mutex = Mutex()
+
+  // The job performing the cache update
+  private var cacheUpdateJob: Job? = null
+
+  /**
+   * Get the size of the cache
+   *
+   * @return The size of the cache
+   */
+  fun getCacheSize(): Int {
+    return cache.size
+  }
+
+  override fun getElevation(
+      coordinates: List<LatLong>,
+      onSuccess: (List<Double>) -> Unit,
+      onFailure: (Exception) -> Unit
+  ) {
+    if (coordinates.isEmpty()) {
+      onSuccess(emptyList())
+      return
+    }
+
+    CoroutineScope(repoDispatcher).launch {
+      mutex.withLock {
+        Log.d(LOG_TAG, "Adding request to requests list")
+        requests.add(ElevationRequest(coordinates, onSuccess, onFailure))
+
+        if (cacheUpdateJob == null || cacheUpdateJob?.isCompleted == true) {
+          launchUpdateOfCache()
+        }
+      }
+    }
+  }
+
+  /**
+   * Launch the update of the cache
+   *
+   * The update will be delayed by some time to allow for multiple requests to be batched together
+   */
+  private fun launchUpdateOfCache() {
+
+    if (requests.isEmpty()) return
+
+    // We need to start a new job to update the cache
+    cacheUpdateJob =
+        CoroutineScope(repoDispatcher).launch {
+          // Delay the update by some amount of time to allow for multiple requests
+          // to be batched together
+          // This is purely empirical and can be adjusted
+          // If we already have enough data for a request, we don't need to wait
+          var numberOfCoordinates: Long
+          var chunks: List<ElevationRequestChunk>
+          mutex.withLock {
+            clearRequestsUsingCache()
+            val parsingResult = parseRequestsIntoChunks()
+            numberOfCoordinates = parsingResult.first
+            chunks = parsingResult.second
+            if (numberOfCoordinates < MAX_COORDINATES_PER_REQUEST) {
+              mutex.unlock()
+              delay(WAITING_DELAY_BEFORE_SENDING_REQUEST_FOR_BATCHING)
+              mutex.lock()
```

Comment:
Actually, nothing in the documentation states to avoid it, and the functions would throw an `IllegalStateException` if there were any issue! But I agree it looks cursed 😂

Reply:
Alright as long as it works 😂